clippy_config/src/conf.rs

```diff
@@ -42,217 +41,205 @@ const DEFAULT_DOC_VALID_IDENTS: &[&str] = &[
-                                                "duplicate field `", stringify!($new_conf),
-                                                "` (provided as `", stringify!($name), "`)"
-                                            ), None, name.span())),
-                                            None => $new_conf = $name.clone(),
-                                        })?
-                                    },
+                                );
+                                diag.emit();
+
+                                if let Some((other_key, _)) = other_value {
+                                    let mut diag = dcx.dcx.struct_span_err(sp, format!("duplicate key in document root"));
+                                    diag.span_note(dcx.make_sp(other_key.span()), "previous definition here");
+                                    diag.emit();
+                                } else {
+                                    continue;
                                 }
-                            }
-                        })*
-                        // ignore contents of the third_party key
-                        Ok(Field::third_party) => drop(map.next_value::<IgnoredAny>())
+                            },
+                        }
+                        break;
                     }
                 }
-                let conf = Conf { $($name: $name.unwrap_or_else(defaults::$name),)* };
-                Ok(TryConf { conf, errors, warnings })
+
+                Self {$($(
+                    $name: $name.unwrap_or_else(
+                        || <$ty as FromDefault<_>>::from_default(first_expr!($($default,)? ()))
+                    ),
+                )?)*}
             }
         }
 
-        pub fn get_configuration_metadata() -> Vec<ClippyConfiguration> {
-            vec![$(
-                ClippyConfiguration {
-                    name: stringify!($name).replace('_', "-"),
-                    default: default_text!(defaults::$name() $(, $default_text)?),
-                    lints: &[$($(stringify!($for_lints)),*)?],
-                    doc: concat!($($doc, '\n',)*),
-                    deprecation_reason: wrap_option!($($dep)?)
-                },
-            )*]
+        #[test]
+        fn check_conf_order() {
+            for [x, y] in ConfField::NAMES[..ConfField::NAMES.len() - 1].array_windows::<2>() {
+                assert!(x <= y, "configuration `{x}` and `{y}` are out of order");
+            }
         }
+
+        #[test]
+        fn check_conf_names() {$(
+            assert_eq!(stringify!($name).replace('_', "-"), $name_str);
+        )*}
     };
 }
 
 define_Conf! {
     /// Which crates to allow absolute paths from
     #[lints(absolute_paths)]
-    absolute_paths_allowed_crates: FxHashSet<String> = FxHashSet::default(),
+    absolute_paths_allowed_crates("absolute-paths-allowed-crates"): FxHashSet<String>,
```

Comment:
I don't really like this separation of the field name and the toml name. What is the motivation for this change?

Reply:
Not being able to use `serde` anymore. Switching to a proc-macro would fix this and some other issues.

Reply:
Ahh okay, it makes sense and with the test you added, it should be ensured that they always match up. I'll take a look at the deserialization soon.